### PR TITLE
Datafeer: Add wizard form

### PR DESCRIPTION
**Pull request by fgravin**
The PR implement the wizard form of the data feeder.
See https://github.com/geonetwork/geonetwork-ui/discussions/70 for the form configuration specifications.

The form is auto generated from the config, the steps are binded with the router state.

2 dependencies are added for specific inputs:

- `ngx-chips` for the autocomplete tagsinput (keyword picker)
- `mydatepicker` for the date

### CODE REVIEW

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        todayBtnTxt: 'Today',
+        markCurrentDay: true,
+        dateFormat: 'dd mmm yyyy',
+        firstDayOfWeek: 'mo',
+        sunHighlight: true,
+        inline: false,
+        showTodayBtn: false,
+        height: '100%',
+        selectorHeight: '100%',
+        monthLabels: monthLabels,
```
Comment (fgravin) [inline apps/datafeeder/src/app/presentation/components/form-field/form-field.component.ts:74]: no way to use labels from `Date.toLocaleDateString` ?

File: apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.css [hunk 2]
```diff
@@ -0,0 +1,12 @@
+:host {
+  display: flex;
+  flex: 1;
+}
+
+.dataset-forms-bkg {
```
Comment (fgravin) [inline apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.css:6]:
```suggestion
.dataset-forms-bg {
```

File: apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.html [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<div class="flex-1 flex dataset-forms-bkg items-center">
+  <div class="ml-64 mr-64 pt-16 pb-20 w-full h-full flex flex-col">
+    <div class="text-5xl font-bold" translate>
+      datafeeder.formsPage.title
+    </div>
+    <div class="pb-4 pt-4">
+      <ui-step-bar
+        [steps]="6"
```
Comment (fgravin) [inline apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.html:8]: it's not hardcoded, it depends on the configuration

File: apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  constructor(
+    private activatedRoute: ActivatedRoute,
+    private router: Router,
+    private logService: LogService
+  ) {}
+
+  ngOnInit(): void {
+    this.routeParamsSub = this.activatedRoute.params.subscribe(
+      ({ id, stepId }) => {
+        this.logService.log(`id: ${id}`)
```
Comment (fgravin) [inline apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts:73]: useless to log that

File: apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        this.logService.log(`stepId: ${stepId}`)
+
+        this.rootId = id
+        this.stepId = Number(stepId)
+      }
+    )
+  }
+
+  handleNextBtnClick() {
+    if (this.stepId === 4) {
```
Comment (fgravin) [inline apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts:83]: what is 4 ? no hardcoded value please

File: apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  handlePreviousBtnClick() {
+    if (this.stepId === 1) {
+      this.router.navigate(['/', this.rootId, 'validation'])
+
+      return
+    }
+
+    this.router.navigate(['/', this.rootId, 'step', --this.stepId])
```
Comment (fgravin) [inline apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts:97]: factorize: add a method `private changeStep(step: number)`

File: libs/ui/src/lib/chips-input/chips-input.component.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  rawChange: BehaviorSubject<Items[]>
+
+  items: Items[] = []
+
+  onChange(event) {
+    this.rawChange.next(this.items)
+  }
+
+  requestAutocompleteItems = (text: string): Observable<any> => {
+    const url = this.url(text)
```
Comment (fgravin) [inline libs/ui/src/lib/chips-input/chips-input.component.ts:30]: why is it a function ?

File: libs/ui/src/lib/datepicker/datepicker.component.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import { Component, Input, OnInit } from '@angular/core'
+import { IMyDateModel, IMyDpOptions } from 'mydatepicker'
+
+@Component({
+  selector: ' ui-datepicker',
+  templateUrl: './datepicker.component.html',
+  styleUrls: ['./datepicker.component.css'],
+})
+export class DatepickerComponent implements OnInit {
+  @Input() options: IMyDpOptions
```
Comment (fgravin) [inline libs/ui/src/lib/datepicker/datepicker.component.ts:10]: no output to emit the date ?

File: libs/ui/src/lib/dropdown-selector/dropdown-selector.component.html [hunk 9]
```diff
@@ -1,5 +1,6 @@
 <div class="flex flex-row items-center relative">
   <label
+    *ngIf="showTitle"
```
Comment (fgravin) [inline libs/ui/src/lib/dropdown-selector/dropdown-selector.component.html:3]:
```suggestion
    *ngIf="title"
```
wouldn't make it ?

File: libs/ui/src/lib/text-area/text-area.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  let component: TextAreaComponent
+  let fixture: ComponentFixture<TextAreaComponent>
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [TextAreaComponent],
+    }).compileComponents()
+  }))
+
+  beforeEach(() => {
```
Comment (fgravin) [inline libs/ui/src/lib/text-area/text-area.component.spec.ts:15]: no test ?

**Review by fgravin — COMMENTED** [review]
I've done a first quick review. Globally the logic has been understood but many things need to be improved:
- the global architecture of dumb/smart component is not masterized. Some code and types about the wizard and form fields should be in the `editor` library
- the code is not tested
- still a lot of hardcoded values
- date localization

File: angular.json [hunk 11]
```diff
@@ -488,5 +488,5 @@
       }
     }
   },
-  "defaultProject": "app-search"
+  "defaultProject": "datafeeder"
```
[the remark this answers is not in the document]
Comment (y-vynohradov) [reply angular.json:491]: okay

File: apps/datafeeder/src/app/app.module.ts [hunk 12]
```diff
@@ -35,14 +36,22 @@ import { AnalysisProgressIllustrationsComponent } from './presentation/component
...
   ],
   imports: [
     BrowserModule,
     AppRoutingModule,
+    I18nModule,
     UiModule,
     HttpClientModule,
-    I18nModule,
-    TranslateModule.forRoot(TRANSLATE_DEFAULT_CONFIG),
+    TranslateModule.forRoot({
```
[the remark this answers is not in the document]
Comment (y-vynohradov) [reply apps/datafeeder/src/app/app.module.ts:48]: It wasn't worked for me. I will chek and change it

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css [hunk 13]
```diff
@@ -0,0 +1,30 @@
+:host {
+  display: flex;
```
[the remark this answers is not in the document]
Comment (y-vynohradov) [reply apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css:2]: I didn't see something simular in tailwindCSS

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+  flex: 1;
+}
+
+ui-text-input::ng-deep input {
+  height: 100%;
+  background: white;
+  color: black;
```
[the remark this answers is not in the document]
Comment (y-vynohradov) [reply apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css:9]: TW can't change color of input that placed in ui-text-input component

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    }).compileComponents()
+  }))
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(FormFieldComponent)
+    component = fixture.componentInstance
+    fixture.detectChanges()
+  })
+
+  it('should create', () => {
```
[the remark this answers is not in the document]
Comment (y-vynohradov) [reply apps/datafeeder/src/app/presentation/components/form-field/form-field.component.spec.ts:21]: Okay

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+export enum FormFieldType {
+  TEXT,
+  TEXT_AREA,
+  CHIPS,
+  DATA_PICKER,
+  DROPDOWN,
+}
+
+const MONTH_OF_THE_YEAR = [
```
[the remark this answers is not in the document]
Comment (y-vynohradov) [reply apps/datafeeder/src/app/presentation/components/form-field/form-field.component.ts:14]: https://github.com/kekeh/mydatepicker
Look at "options attribute"

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.ts (hunk 1 above)
Comment (y-vynohradov) [reply apps/datafeeder/src/app/presentation/components/form-field/form-field.component.ts:74]: yes

File: apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts (hunk 6 above)
Comment (y-vynohradov) [reply apps/datafeeder/src/app/presentation/pages/forms-page/forms-page.component.ts:97]: okay

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css (hunk 13 above)
[the remark this answers is not in the document]
Comment (fgravin) [reply apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css:2]:
```html
<div class="flex h-full">
  <div class="flex-1">
```
etc..

File: apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css (hunk 14 above)
[the remark this answers is not in the document]
Comment (fgravin) [reply apps/datafeeder/src/app/presentation/components/form-field/form-field.component.css:9]: Ok I understand, we are not sure to keep TW in the project so I won't ask more here but the correct approach would have been to update `text-input` so it uses css variables to define the background and color ...

File: libs/editor/src/lib/components/wizard/wizard.component.html [hunk 17]
```diff
@@ -0,0 +1,16 @@
+<div class="pl-5 pt-24 flex flex-1 flex-col justify-between">
```
Comment (fgravin) [inline libs/editor/src/lib/components/wizard/wizard.component.html:1]:
```html
<div class="pl-5 pt-24 flex flex-1 flex-col justify-between">
```
and
```css
:host {
  flex: 1;
  display: flex;
  flex-direction: column;
}
```
are not redundant ?

File: apps/datafeeder/src/app/app.module.ts [hunk 18]
```diff
@@ -18,7 +18,7 @@ import { UploadDataBackgroundComponent } from './presentation/components/svg/upl
 import { UploadDataIllustrationComponent } from './presentation/components/svg/upload-data-illustration/upload-data-illustration.component'
 import { AnalysisProgressIllustrationsComponent } from './presentation/components/svg/analysis-progress-illustrations/analysis-progress-illustrations.component'
 import { FormsPageComponent } from './presentation/pages/forms-page/forms-page.component'
-import { FormFieldComponent } from './presentation/components/form-field/form-field.component'
+import { EditorModule } from '../../../../libs/editor/src'
```
Comment (fgravin) [inline apps/datafeeder/src/app/app.module.ts:21]: alias

**Review by fgravin — COMMENTED** [review]
Thanks for the refactoring.
I still don't see the icons

On page loading, there is an error
```
core.js:4197 ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'undefined'. Current value: '5'.
    at throwErrorIfNoChangesMode (core.js:5468)
    at bindingUpdated (core.js:13013)
    at Module.ɵɵproperty (core.js:13808)
    at FormsPageComponent_Template (forms-page.component.html:7)
    at executeTemplate (core.js:7303)
    at refreshView (core.js:7172)
    at refreshComponent (core.js:8326)
    at refreshChildComponents (core.js:6965)
    at refreshView (core.js:7222)
    at refreshEmbeddedViews (core.js:8280)
```

**Comment by fgravin** [comment]
CI still does'nt pass ...
Still formating issues, please fix your node version or the way you run prettier cause this pollute the commit history. Your code should always be correctly formatted, and the test should always pass

**Comment by fgravin** [comment]
I will merge it but several things are not ok
- no test in UI components
- local storage logic is not good (save a dictionnary of batches)
- wizard persistent service might be a singelton binded to a specific id, be sure it can handle several batches
- keyword picker should not use http directly but the open api service `registries`

**Pull request merged**